zerver/lib/users.py

```diff
@@ -575,6 +774,13 @@ def get_raw_user_data(
 
     result = {}
     for row in user_dicts:
+        if row["delivery_email"] is None:
+            # delivery_email field is None only for inaccessible users.
+            # We already have the required data for them in row object,
+            # so we can just add it to result and continue.
+            result[row["user_id"]] = row
```

Comment:
This check using `delivery_email` can be confusing because we actually pass it as `None` for users who cannot access email, but it is done later in the code in `format_user_row` and here it always includes real email.

Reply:
Hmm, this feels really messy. I think a potentially good strategy would be a preparatory refactoring PR that:

* Renames `get_raw_user_data` to `get_users_for_api`
* Creates a `TypedDict` called, say,  `RawUserDict` for the return type of `user_profile_to_user_row` and `get_realm_user_dicts`; and another, `APIUserDict` for the type that `format_user_row` returns, and another for the type that `get_realm_user_dicts` returned.

And then the new `get_user_dicts_in_realm` could return a `Union[APIUserDict, RawUserDict]`, and this block is just storing in `result` a `APIUserDict` if that's what we got back from `get_user_dicts_in_realm`.

Or maybe it'd be better to change `get_user_dicts_in_realm` to return a 2-tuple, one that's a list of `RawUserDict`s and the other than's a list of `APIUserDict`s? But in any case, I think defining types here would allow us to not do this hacky `delivery_email` check.

Reply:
This function now reads really nicely, thanks!